### PR TITLE
Update handling for projectList based WindowExecs to handle window function of window function [databricks] 

### LINKED ISSUE #6531: [BUG] window function of window function queries fail on Databricks 10.4
**Describe the bug**
When running specific window queries on Databricks environments, certain references to child expressions of the window function are actually not provided in the input window expressions when creating the GpuWindowExec. This is different from Apache Spark, as these references are usually provided. This causes the query to throw an exception in the Databricks environment.

This is because `GpuWindowExec.splitAndDedup` makes the following assumption which is not correct in Databricks:
```
This assumes that there is not a window function of another window function, like
`LAG(SUM(a), 2)` which appears to be something all distros split apart into separate
window operations, so we are good.
```

**Steps/Code to reproduce bug**
Python reproduce for Databricks 10.4

```python
from pyspark.sql.types import *

spark.conf.set("spark.rapids.sql.enabled", "true")
spark.conf.set("spark.sql.adaptive.enabled", "false")

data1 = [(1, 2), (2, 3), (3, 4)]
data2 = [(1, 3), (2, 4), (3, 5)]

schema1 = StructType()\
        .add("a_1", LongType())\
        .add("c", LongType())
    
schema2 = StructType()\
        .add("a_2", LongType())\
        .add("b", LongType())
    
spark.createDataFrame(spark.sparkContext.parallelize(data1), schema1).createOrReplaceTempView("agg")
spark.createDataFrame(spark.sparkContext.parallelize(data2), schema2).createOrReplaceTempView("part")

df = spark.sql("select b, sum(c) as sum_c, sum(c)/sum(sum(c)) over (partition by b) as r_c from agg, part where a_1 = a_2 group by b order by b, r_c")
df.collect()
```
Exception:

```
Py4JJavaError: An error occurred while calling o626.collectToPython.
: java.lang.RuntimeException: Couldn't find _w3#1123L in [b#1113L,sum_c#1116L,_gpu_w2#1143L]
	at scala.sys.package$.error(package.scala:30)
	at com.nvidia.spark.rapids.GpuBindReferences$$anonfun$1.applyOrElse(GpuBoundAttribute.scala:87)
	at com.nvidia.spark.rapids.GpuBindReferences$$anonfun$1.applyOrElse(GpuBoundAttribute.scala:81)
	at scala.PartialFunction$OrElse.applyOrElse(PartialFunction.scala:175)
	at scala.PartialFunction$Lifted.apply(PartialFunction.scala:228)
	at scala.PartialFunction$Lifted.apply(PartialFunction.scala:224)
	at com.nvidia.spark.rapids.GpuBindReferences$.transformNoRecursionOnReplacement(GpuBoundAttribute.scala:67)
	at com.nvidia.spark.rapids.GpuBindReferences$.$anonfun$transformNoRecursionOnReplacement$1(GpuBoundAttribute.scala:69)
	at org.apache.spark.sql.catalyst.trees.UnaryLike.mapChildren(TreeNode.scala:1226)
	at org.apache.spark.sql.catalyst.trees.UnaryLike.mapChildren$(TreeNode.scala:1225)
	at org.apache.spark.sql.catalyst.expressions.UnaryExpression.mapChildren(Expression.scala:607)
	at com.nvidia.spark.rapids.GpuBindReferences$.transformNoRecursionOnReplacement(GpuBoundAttribute.scala:69)
	at com.nvidia.spark.rapids.GpuBindReferences$.$anonfun$transformNoRecursionOnReplacement$1(GpuBoundAttribute.scala:69)
	at org.apache.spark.sql.catalyst.trees.BinaryLike.mapChildren(TreeNode.scala:1253)
	at org.apache.spark.sql.catalyst.trees.BinaryLike.mapChildren$(TreeNode.scala:1252)
	at org.apache.spark.sql.catalyst.expressions.BinaryExpression.mapChildren(Expression.scala:699)
	at com.nvidia.spark.rapids.GpuBindReferences$.transformNoRecursionOnReplacement(GpuBoundAttribute.scala:69)
	at com.nvidia.spark.rapids.GpuBindReferences$.$anonfun$transformNoRecursionOnReplacement$1(GpuBoundAttribute.scala:69)
	at org.apache.spark.sql.catalyst.trees.UnaryLike.mapChildren(TreeNode.scala:1226)
	at org.apache.spark.sql.catalyst.trees.UnaryLike.mapChildren$(TreeNode.scala:1225)
	at org.apache.spark.sql.catalyst.expressions.UnaryExpression.mapChildren(Expression.scala:607)
	at com.nvidia.spark.rapids.GpuBindReferences$.transformNoRecursionOnReplacement(GpuBoundAttribute.scala:69)
	at com.nvidia.spark.rapids.GpuBindReferences$.bindRefInternal(GpuBoundAttribute.scala:93)
	at com.nvidia.spark.rapids.GpuBindReferences$.bindGpuReference(GpuBoundAttribute.scala:101)
	at com.nvidia.spark.rapids.GpuBindReferences$.$anonfun$bindGpuReferences$1(GpuBoundAttribute.scala:111)
	at scala.collection.immutable.List.map(List.scala:297)
	at com.nvidia.spark.rapids.GpuBindReferences$.bindGpuReferences(GpuBoundAttribute.scala:111)
	at com.nvidia.spark.rapids.GpuProjectExec.doExecuteColumnar(basicPhysicalOperators.scala:144)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeColumnar$1(SparkPlan.scala:251)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:268)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:165)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:264)
	at org.apache.spark.sql.execution.SparkPlan.executeColumnar(SparkPlan.scala:247)
	at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase.inputBatchRDD$lzycompute(GpuShuffleExchangeExecBase.scala:201)
	at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase.inputBatchRDD(GpuShuffleExchangeExecBase.scala:201)
	at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase.shuffleDependencyColumnar$lzycompute(GpuShuffleExchangeExecBase.scala:211)
	at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase.shuffleDependencyColumnar(GpuShuffleExchangeExecBase.scala:209)
	at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase.$anonfun$doExecuteColumnar$1(GpuShuffleExchangeExecBase.scala:235)
	at com.nvidia.spark.rapids.shims.Spark320PlusShims.attachTreeIfSupported(Spark320PlusShims.scala:291)
	at com.nvidia.spark.rapids.shims.Spark320PlusShims.attachTreeIfSupported$(Spark320PlusShims.scala:286)
	at com.nvidia.spark.rapids.shims.SparkShimImpl$.attachTreeIfSupported(SparkShims.scala:41)
	at org.apache.spark.sql.rapids.execution.GpuShuffleExchangeExecBase.doExecuteColumnar(GpuShuffleExchangeExecBase.scala:232)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeColumnar$1(SparkPlan.scala:251)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:268)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:165)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:264)
	at org.apache.spark.sql.execution.SparkPlan.executeColumnar(SparkPlan.scala:247)
	at com.nvidia.spark.rapids.GpuShuffleCoalesceExec.doExecuteColumnar(GpuShuffleCoalesceExec.scala:66)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeColumnar$1(SparkPlan.scala:251)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:268)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:165)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:264)
	at org.apache.spark.sql.execution.SparkPlan.executeColumnar(SparkPlan.scala:247)
	at com.nvidia.spark.rapids.GpuSortExec.doExecuteColumnar(GpuSortExec.scala:128)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeColumnar$1(SparkPlan.scala:251)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:268)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:165)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:264)
	at org.apache.spark.sql.execution.SparkPlan.executeColumnar(SparkPlan.scala:247)
	at com.nvidia.spark.rapids.GpuColumnarToRowExec.doExecute(GpuColumnarToRowExec.scala:320)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$execute$1(SparkPlan.scala:224)
	at org.apache.spark.sql.execution.SparkPlan.$anonfun$executeQuery$1(SparkPlan.scala:268)
	at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:165)
	at org.apache.spark.sql.execution.SparkPlan.executeQuery(SparkPlan.scala:264)
	at org.apache.spark.sql.execution.SparkPlan.execute(SparkPlan.scala:220)
	at org.apache.spark.sql.execution.collect.Collector$.collect(Collector.scala:78)
	at org.apache.spark.sql.execution.collect.Collector$.collect(Collector.scala:87)
	at org.apache.spark.sql.execution.collect.InternalRowFormat$.collect(cachedSparkResults.scala:75)
	at org.apache.spark.sql.execution.collect.InternalRowFormat$.collect(cachedSparkResults.scala:62)
	at org.apache.spark.sql.execution.ResultCacheManager.collectResult$1(ResultCacheManager.scala:575)
	at org.apache.spark.sql.execution.ResultCacheManager.computeResult(ResultCacheManager.scala:582)
	at org.apache.spark.sql.execution.ResultCacheManager.$anonfun$getOrComputeResultInternal$1(ResultCacheManager.scala:528)
	at scala.Option.getOrElse(Option.scala:189)
	at org.apache.spark.sql.execution.ResultCacheManager.getOrComputeResultInternal(ResultCacheManager.scala:527)
	at org.apache.spark.sql.execution.ResultCacheManager.getOrComputeResult(ResultCacheManager.scala:424)
	at org.apache.spark.sql.execution.ResultCacheManager.getOrComputeResult(ResultCacheManager.scala:403)
	at org.apache.spark.sql.execution.SparkPlan.executeCollectResult(SparkPlan.scala:422)
	at org.apache.spark.sql.Dataset.$anonfun$collectToPython$1(Dataset.scala:3739)
	at org.apache.spark.sql.Dataset.$anonfun$withAction$1(Dataset.scala:3951)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withCustomExecutionEnv$8(SQLExecution.scala:239)
	at org.apache.spark.sql.execution.SQLExecution$.withSQLConfPropagated(SQLExecution.scala:386)
	at org.apache.spark.sql.execution.SQLExecution$.$anonfun$withCustomExecutionEnv$1(SQLExecution.scala:186)
	at org.apache.spark.sql.SparkSession.withActive(SparkSession.scala:968)
	at org.apache.spark.sql.execution.SQLExecution$.withCustomExecutionEnv(SQLExecution.scala:141)
	at org.apache.spark.sql.execution.SQLExecution$.withNewExecutionId(SQLExecution.scala:336)
	at org.apache.spark.sql.Dataset.withAction(Dataset.scala:3949)
	at org.apache.spark.sql.Dataset.collectToPython(Dataset.scala:3737)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)
	at py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:380)
	at py4j.Gateway.invoke(Gateway.java:295)
	at py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)
	at py4j.commands.CallCommand.execute(CallCommand.java:79)
	at py4j.GatewayConnection.run(GatewayConnection.java:251)
	at java.lang.Thread.run(Thread.java:748)
```

### PR DESCRIPTION
**Pull request by NVnavkumar**
Fixes #6531.

This updates the logic that splits and dedups window expressions for GpuWindowExec to handle the Databricks case to include input fields into the WindowExec computation while maintaining the correct output of the WindowExec. To explain this further:

1. Databricks implementation of WindowExec uses the `projectList` method instead of the `windowExpression` method that Apache Spark uses. (Effectively it combines a ProjectExec and a WindowExec together, since `projectList` is originally a ProjectExec method)
2. In that case we mark this situation as `resultColumnsOnly` which means that we only want the output columns in the result of the WindowExec (makes sense since the WindowExec is performing double duty as a window computation and a projection)
3. Initially, we were just taking originally the output of `projectList` into `GpuWindowExec.splitAndDedup` which by definition would not include all the fields necessary in the window computation (this is due to Apache Spark returning everything in the output of the WindowExec, and would add an additional ProjectExec if that was needed)
4. In Databricks, this would throw an exception in cases where the window function needed input fields that were not in the output projection since we were only using `projectList` which only includes output columns.
5. Therefore, we need to add these input fields back for the pre project stage and the window operation stage, but exclude them from the post project since we only want the result output columns only in Databricks
6. Since the post project stage will then only include result output columns ultimately, the logic previously in place can add the necessary GpuProjectExec required to output only the correct columns.

### CODE REVIEW

**Comment by NVnavkumar** [comment]
build

**Comment by NVnavkumar** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExec.scala [hunk 1]
```diff
@@ -217,6 +225,7 @@ class GpuWindowExecMeta(windowExec: WindowExec,
     } catch {
       case _: NoSuchMethodException =>
         resultColumnsOnly = true
+        val methodNames = windowExec.getClass.getMethods().map(_.getName).toList
```
Comment (mythrocks) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExec.scala:228]: Where is `methodNames` used?

Comment (NVnavkumar) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExec.scala:228]: oops, old debug code, that should be removed.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExec.scala [hunk 2]
```diff
@@ -231,7 +239,7 @@ class GpuWindowExecMeta(windowExec: WindowExec,
   override def getResultColumnsOnly: Boolean = resultColumnsOnly
 }
 
-object GpuWindowExec extends Arm {
+object GpuWindowExec extends Arm with Logging {
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExec.scala:242]: nit: I don't think you need this logging here.

**Review by revans2 — DISMISSED** [review]
Just some nits.

**Review by mythrocks — DISMISSED** [review]
A couple of minor nitpicks aside, this LGTM! Excellent stuff.

File: integration_tests/src/main/python/window_function_test.py [hunk 3]
```diff
@@ -907,6 +907,33 @@ def test_running_window_function_exec_for_all_aggs():
...
         ''')
 
+# Test the Databricks WindowExec which combines a WindowExec with a ProjectExec and provides the output
+# fields that we need to handle with an extra GpuProjectExec and we need the input expressions to compute
+# a window function of another window function case
+@ignore_order(local=True)
+@pytest.mark.parametrize('data_gen', integral_gens, ids=idfn)
+def test_join_sum_window_of_window(data_gen):
+    conf = {'spark.rapids.sql.batchSizeBytes': '100',
+            'spark.rapids.sql.explain': 'NONE'}
```
[the remark this answers is not in the document]
Comment (NVnavkumar) [reply integration_tests/src/main/python/window_function_test.py:917]: debugging reasons. removing this config.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExec.scala (hunk 2 above)
Comment (NVnavkumar) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuWindowExec.scala:242]: thought I caught that one, will remove.

**Review by revans2 — DISMISSED** [review]

**Review by revans2 — APPROVED** [review]

**Comment by NVnavkumar** [comment]
build

**Comment by sameerz** [comment]
pre-merge blocked on https://github.com/rapidsai/cudf/pull/11699

**Comment by sameerz** [comment]
build

**Comment by NVnavkumar** [comment]
build

**Comment by sameerz** [comment]
build

**Comment by revans2** [comment]
The test failure looks related to UCX issues on databricks. Not 100% sure why that is happening. Will retry the build just in case.

**Comment by revans2** [comment]
build

**Comment by NVnavkumar** [comment]
build

**Pull request merged**